### PR TITLE
Add guard against half support

### PR DESCRIPTION
**Pull request by miscco**
This mirrors our current escape hatch against `__nv_bfloat16` to also allow the user to disable half support

Addresses: #1727

### CODE REVIEW

File: cub/cub/device/dispatch/dispatch_histogram.cuh [hunk 1]
```diff
@@ -773,7 +761,9 @@ public:
     _CCCL_HOST_DEVICE _CCCL_FORCEINLINE int
     ComputeBin(T sample, T min_level, ScaleT scale, ::cuda::std::false_type /* is_fp */)
     {
-      return static_cast<int>(((sample - min_level) * scale.fraction.bins) / scale.fraction.range);
+      return static_cast<int>(::cuda::std::divides<T>{}(
+        ::cuda::std::multiplies<T>{}(::cuda::std::minus<T>{}(sample, min_level), scale.fraction.bins),
+        scale.fraction.range));
```
Comment (gevtushenko) [inline cub/cub/device/dispatch/dispatch_histogram.cuh:766]: critical: this yields different result in presence of integer promotion. I don't think we can make this change:
```cpp
  char sample = 100;
  char min_level = 0;
  char max_level = 120;
  CommonT bins = 120;
  CommonT range = max_level - min_level;
  std::cout << OldComputeBin(sample, min_level, bins, range) << std::endl; // 100
  std::cout << NewComputeBin(sample, min_level, bins, range) << std::endl; // 0
```

File: cub/cub/util_type.cuh [hunk 2]
```diff
@@ -46,24 +46,24 @@
...
 
-#if !_NVHPC_CUDA
+#if !defined(_CCCL_CUDA_COMPILER_NVHPC) && !defined(CUB_DISABLE_FP16_SUPPORT)
 #  include <cuda_fp16.h>
-#endif
+#endif // !_CCCL_CUDA_COMPILER_NVHPC && !CUB_DISABLE_FP16_SUPPORT
+
+#if !defined(CUB_DISABLE_BF16_SUPPORT) && defined(CUB_DISABLE_FP16_SUPPORT)
+#  define CUB_DISABLE_BF16_SUPPORT
+#endif // !CUB_DISABLE_BF16_SUPPORT && CUB_DISABLE_FP16_SUPPORT
```
Comment (gevtushenko) [inline cub/cub/util_type.cuh:55]: suggestion: it's not exactly trivial that bf16 includes fp16, and if fp16 is disabled we also have to disable bf16. Could you please add a comment? Also, we might want to document this for users:
```cpp
#ifdef DOXYGEN_SHOULD_SKIP_THIS

/**
 * @def CCCL_DISABLE_FP16_SUPPORT
 *
 * Has to disable BF16 support as well, etc.
 */
#  define CCCL_DISABLE_FP16_SUPPORT

#endif
```

**Review by gevtushenko — CHANGES_REQUESTED** [review]
Left a few comments. Please, let me know when CTK 12.1 can build our tests and I'll have another look.

File: cub/cmake/CubHeaderTesting.cmake [hunk 3]
```diff
@@ -48,3 +48,6 @@ cub_add_header_test(base "${header_definitions}")
 
 list(APPEND header_definitions "CUB_DISABLE_BF16_SUPPORT")
 cub_add_header_test(bf16 "${header_definitions}")
+
+list(APPEND header_definitions "CUB_DISABLE_FP16_SUPPORT")
+cub_add_header_test(half "${header_definitions}")
```
[the remark this answers is not in the document]
Comment (miscco) [reply cub/cmake/CubHeaderTesting.cmake:53]: I changed it so that we are more explicit

**Comment by miscco** [comment]
@gevtushenko I have reworked the approach and ditched the `cuda::std::meow` functors.

It now directly utilizes the builtin functions when possible and falls back to float when needed.

I also went ahead and started a centralization of the detection so that we do not duplicate all that logic

File: cub/cub/util_type.cuh [hunk 4]
```diff
@@ -43,28 +43,6 @@
...
-
-#if !_NVHPC_CUDA && !defined(CUB_DISABLE_BF16_SUPPORT)
-#  include <cuda_bf16.h>
-// cuda_fp8.h transitively includes cuda_fp16.h, so we have to include the header under !CUB_DISABLE_BF16_SUPPORT
-#  if _CCCL_CUDACC_VER >= 1108000
-// cuda_fp8.h resets default for C4127, so we have to guard the inclusion
-#    if defined(_CCCL_COMPILER_MSVC)
-#      pragma warning(push)
-#    endif
-#    include <cuda_fp8.h>
```
Comment (gevtushenko) [inline cub/cub/util_type.cuh:61]: critical: the removed code used to include fp8 header. Code below relies on `__CUDA_FP8_TYPES_EXIST__`. The change removed fp8 include along with the fp8 types support. This now doesn't compile:

```cpp
#include <cub/block/block_radix_sort.cuh>
#include <cuda_fp8.h>

__global__ void kernel() {
    __nv_fp8_e4m3 data[3];

    using block_sort_t = cub::BlockRadixSort<__nv_fp8_e4m3, 256, 3>;
    using temp_storage_t = block_sort_t::TempStorage;

    __shared__ temp_storage_t temp_storage;
    block_sort_t(temp_storage).Sort(data);
}
```

**Review by gevtushenko — CHANGES_REQUESTED** [review]
Thank you for addressing the issues! There are a few more breaking changes. Please, let me know when those are addressed and I'll have another look.

File: cub/cmake/CubHeaderTesting.cmake [hunk 5]
```diff
@@ -46,5 +46,14 @@ set(header_definitions
   "CUB_WRAPPED_NAMESPACE=wrapped_cub")
 cub_add_header_test(base "${header_definitions}")
 
-list(APPEND header_definitions "CUB_DISABLE_BF16_SUPPORT")
+set(header_definitions
+  "THRUST_WRAPPED_NAMESPACE=wrapped_thrust"
+  "CUB_WRAPPED_NAMESPACE=wrapped_cub"
+  "CCCL_DISABLE_BF16_SUPPORT")
```
[the remark this answers is not in the document]
Comment (gevtushenko) [reply cub/cmake/CubHeaderTesting.cmake:52]: I think we should keep testing with `CUB_DISABLE_BF16_SUPPORT` to make sure we do not break it

**Review by gevtushenko — APPROVED** [review]

**Pull request merged**